File: Accupay.Data/Entities/PayPeriod.cs
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using AccuPay.Data.Services;
using AccuPay.Utilities.Extensions;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("payperiod")]
    public class PayPeriod : BaseEntity, IPayPeriod
    {
        private const int IsJanuary = 1;
        private const int IsDecember = 12;

        public const int FirstHalfValue = 1;
        public const int EndOftheMonthValue = 0;

        public int? OrganizationID { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        public int? CreatedBy { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime? LastUpd { get; set; }

        public int? LastUpdBy { get; set; }

        [Column("TotalGrossSalary")]
        public int? PayFrequencyID { get; set; }

        public DateTime PayFromDate { get; set; }

        public DateTime PayToDate { get; set; }

        public int Month { get; set; }

        public int Year { get; set; }

        public int Half { get; set; }

        public int OrdinalValue { get; set; }

        public PayPeriodStatus Status { get; set; }

        public bool SSSWeeklyContribSched { get; set; }
        public bool PhHWeeklyContribSched { get; set; }
        public bool HDMFWeeklyContribSched { get; set; }
        public bool WTaxWeeklyContribSched { get; set; }

        public bool SSSWeeklyAgentContribSched { get; set; }
        public bool PhHWeeklyAgentContribSched { get; set; }
        public bool HDMFWeeklyAgentContribSched { get; set; }
        public bool WTaxWeeklyAgentContribSched { get; set; }
        public bool IsClosed { get; set; }

        public virtual ICollection<Paystub> Paystubs { get; set; }

        public bool IsSemiMonthly => PayFrequencyID == PayrollTools.PayFrequencySemiMonthlyId;
        public bool IsWeekly => PayFrequencyID == PayrollTools.PayFrequencyWeeklyId;
        public bool IsFirstHalf => Half == FirstHalfValue;
        public bool IsEndOfTheMonth => Half == EndOftheMonthValue;

        public bool IsBetween(DateTime date)
        {
            date = date.ToMinimumHourValue();
            return date >= PayFromDate && date <= PayToDate;
        }

        public bool IsFirstPayPeriodOfTheYear => IsFirstHalf && Month == IsJanuary;
        public bool IsLastPayPeriodOfTheYear => IsEndOfTheMonth && Month == IsDecember;

        public static PayPeriod NewPayPeriod(int organizationId, int month, int year, bool isFirstHalf, PolicyHelper policy)
        {
            var ordinalValue = month * 2;

            var firstHalfDaySpan = policy.DefaultFirstHalfDaysSpan();
            var endOfTheMonthDaySpan = policy.DefaultEndOfTheMonthDaysSpan();

            DateTime payFromDate, payToDate;

            if (isFirstHalf)
            {
                ordinalValue--;
                payFromDate = firstHalfDaySpan.From.GetDate(month: month, year: year);
                payToDate = firstHalfDaySpan.To.GetDate(month: month, year: year);
            }
            else
            {
                payFromDate = endOfTheMonthDaySpan.From.GetDate(month: month, year: year);
                payToDate = endOfTheMonthDaySpan.To.GetDate(month: month, year: year);
            }

            return new PayPeriod()
            {
                OrganizationID = organizationId,
                PayFrequencyID = PayrollTools.PayFrequencySemiMonthlyId,
                Month = month,
                Year = year,
                Half = isFirstHalf ? FirstHalfValue : EndOftheMonthValue,
                Status = PayPeriodStatus.Pending,
                IsClosed = false,
                OrdinalValue = ordinalValue,

                PayFromDate = payFromDate,
                PayToDate = payToDate
            };
        }
    }
}

File: AccuPay/Domain/Paystubs/PayPeriod.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema
Imports AccuPay.Data

Namespace Global.AccuPay.Entity

    <Table("payperiod")>
    Public Class PayPeriod
        Implements IPayPeriod

        Private Const IsJanuary As Integer = 1
        Private Const IsDecember As Integer = 12

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer? Implements IPayPeriod.RowID

        Public Property OrganizationID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        <Column("TotalGrossSalary")>
        Public Property PayFrequencyID As Integer?

        Public Property PayFromDate As Date Implements IPayPeriod.PayFromDate

        Public Property PayToDate As Date Implements IPayPeriod.PayToDate

        Public Property Month As Integer

        Public Property Year As Integer

        Public Property Half As Integer

        Public Property OrdinalValue As Integer

        Public Property SSSWeeklyContribSched As Boolean
        Public Property PhHWeeklyContribSched As Boolean
        Public Property HDMFWeeklyContribSched As Boolean
        Public Property WTaxWeeklyContribSched As Boolean

        Public Property SSSWeeklyAgentContribSched As Boolean
        Public Property PhHWeeklyAgentContribSched As Boolean
        Public Property HDMFWeeklyAgentContribSched As Boolean
        Public Property WTaxWeeklyAgentContribSched As Boolean
        Public Property IsClosed As Boolean

        Public ReadOnly Property IsSemiMonthly As Boolean
            Get
                Return PayFrequencyID.Value = Data.Helpers.PayrollTools.PayFrequencySemiMonthlyId
            End Get
        End Property

        Public ReadOnly Property IsWeekly As Boolean
            Get
                Return PayFrequencyID.Value = Data.Helpers.PayrollTools.PayFrequencyWeeklyId
            End Get
        End Property

        Public ReadOnly Property IsFirstHalf As Boolean
            Get
                Return Half = 1
            End Get
        End Property

        Public ReadOnly Property IsEndOfTheMonth As Boolean
            Get
                Return Half = 0
            End Get
        End Property

        Public ReadOnly Property IsBetween([date] As Date) As Boolean
            Get
                Return [date] >= PayFromDate AndAlso [date] <= PayToDate
            End Get
        End Property

        Public ReadOnly Property IsFirstPayPeriodOfTheYear As Boolean
            Get
                Return IsFirstHalf AndAlso Month = IsJanuary
            End Get
        End Property

        Public ReadOnly Property IsLastPayPeriodOfTheYear As Boolean
            Get
                Return IsEndOfTheMonth AndAlso Month = IsDecember
            End Get
        End Property

    End Class

End Namespace